### PR TITLE
take scaling action when alarm is received

### PR DESCRIPTION
**Pull request by boyang9527**
[#119505369]

- server to handle triggers from eventgen
- add get/set instance number in cf client
- actively refresh token in cf client
- add GetAppPolicy in shared db module
- change test case of RetrievePolicys in db module
- share models package
- change eventgen to use type metric/instanceMetrics  in shared models

not in scope

- scaling engine configuration
- scaling engine main
- handling cool down

### CODE REVIEW

File: src/cf/app.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+)
+
+const (
+	TokenTypeBearer = "bearer"
+	PathAppSummary  = "/v2/apps/{appid}/summary"
+	PathAppUpdate   = "/v2/apps/{appid}"
+)
+
+func (c *cfClient) GetAppInstances(appId string) (int, error) {
+	url := c.conf.Api + strings.Replace(PathAppSummary, "{appid}", appId, 1)
```
Comment (fraenkel) [inline src/cf/app.go]: Let us use the PathAppUpdate instead. Summary is more expensive and unnecessary to get the instance count.

File: src/cf/client.go [hunk 2]
```diff
@@ -186,6 +199,21 @@ func (c *cfClient) GetTokens() Tokens {
 	return c.tokens
 }
 
+func (c *cfClient) GetTokensWithRefresh() Tokens {
+	if c.isTokenToBeExpired() {
```
Comment (fraenkel) [inline src/cf/client.go]: Please don't do double-lock checking. Just lock and check.

File: src/cf/config.go [hunk 3]
```diff
@@ -18,6 +20,18 @@ func (conf *CfConfig) Validate() error {
 		return fmt.Errorf("Configuration error: cf api is empty")
 	}
 
+	apiUrl, err := url.Parse(conf.Api)
+	if err != nil {
+		return fmt.Errorf("Configuration error: cf api is invalid")
+	}
+	if apiUrl.Scheme == "" {
+		apiUrl.Scheme = "https"
```
Comment (fraenkel) [inline src/cf/config.go]: We should require the scheme. Having them guess our default isn't great.

File: src/cf/config_test.go [hunk 4]
```diff
@@ -33,7 +33,28 @@ var _ = Describe("Config", func() {
...
+			BeforeEach(func() {
+				conf.Api = "a.com%"
+			})
+
+			It("should error", func() {
+				Expect(err).To(MatchError("Configuration error: cf api is invalid"))
+			})
+		})
+
+		Context("when api valid but not regulated", func() {
```
Comment (fraenkel) [inline src/cf/config_test.go]: but ends with a /

File: src/engine/server/scaling_handler.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	var newInstances int
+	newInstances, err = h.Scale(trigger)
+	if err != nil {
+		h.logger.Error("handle-scale-perform-scaling-action", err, lager.Data{"trigger": trigger})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Internal-server-error",
+			Message: "Error taking scaling action"})
+		return
+	}
+	handlers.WriteJSONResponse(w, http.StatusCreated, models.AppProperties{Instances: newInstances})
```
Comment (fraenkel) [inline src/engine/server/scaling_handler.go]: Is this really a 201? Shouldn't this be a 200? Since it can be either created or updated.

File: src/engine/server/scaling_handler_test.go [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+		newInstances int
+		adjustment   string
+		trigger      *models.Trigger
+		buffer       *gbytes.Buffer
+	)
+
+	BeforeEach(func() {
+		cfc = &fakes.FakeCfClient{}
+
+		logger := lager.NewLogger("scaling-handler-test")
```
Comment (fraenkel) [inline src/engine/server/scaling_handler_test.go]: Use lagertest.NewTestLogger

File: src/engine/server/server.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	"db"
+
+	"code.cloudfoundry.org/lager"
+	"github.com/gorilla/mux"
+	"github.com/tedsuo/ifrit"
+	"github.com/tedsuo/ifrit/http_server"
+)
+
+const (
+	PathScalingActions = "/v1/scaling_actions"
```
Comment (fraenkel) [inline src/engine/server/server.go]: Do we want to call this /v1/scale? or /v1/scalingengine/scale?

File: src/eventgenerator/aggregator/metricPoller_test.go [hunk 8]
```diff
@@ -3,8 +3,8 @@ package aggregator_test
 import (
 	. "eventgenerator/aggregator"
 	. "eventgenerator/appmetric"
-	. "metricscollector/metrics"
-	. "metricscollector/server"
+	// . "metricscollector/server"
```
Comment (fraenkel) [inline src/eventgenerator/aggregator/metricPoller_test.go]: Why is this here?

File: src/metricscollector/cmd/metricscollector/metricscollector_suite_test.go [hunk 9]
```diff
@@ -106,12 +106,11 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	cfg.Db.MetricsDbUrl = os.Getenv("DBURL")
 	cfg.Db.PolicyDbUrl = os.Getenv("DBURL")
 
-	cfg.Collector.PollInterval = 10 * time.Second
-	cfg.Collector.RefreshInterval = 30 * time.Second
+	cfg.Collector.PollInterval = 10
```
Comment (fraenkel) [inline src/metricscollector/cmd/metricscollector/metricscollector_suite_test.go]: Do you really mean 10 ns?

File: src/metricscollector/config/config.go [hunk 10]
```diff
@@ -4,17 +4,16 @@ import (
...
 
 	"cf"
 )
 
 const (
-	DefaultLoggingLevel                  = "info"
-	DefaultRefreshInterval time.Duration = 60 * time.Second
-	DefaultPollInterval    time.Duration = 30 * time.Second
+	DefaultLoggingLevel    = "info"
+	DefaultRefreshInterval = 60
```
Comment (fraenkel) [inline src/metricscollector/config/config.go]: Why are we switching to int from Duration?  Duration gives us more flexibility.

File: src/cf/app.go (hunk 1 above)
Comment (boyang9527) [inline src/cf/app.go]: sure, will change to /v2/apps/{appid}

File: src/cf/client.go (hunk 2 above)
Comment (boyang9527) [inline src/cf/client.go]: this is used to avoid another  refresh if other thread has already refreshed the token and changed the expire time.

We can remove the outer check while this will have more overhead since  no matter the token is to be expired or not we need lock/unlock

File: src/cf/config.go (hunk 3 above)
Comment (boyang9527) [inline src/cf/config.go]: I will change it to fail the validation if there is no scheme

File: src/cf/config_test.go (hunk 4 above)
Comment (boyang9527) [inline src/cf/config_test.go]: sure will change to make the work more precise

File: src/engine/server/scaling_handler.go (hunk 5 above)
Comment (boyang9527) [inline src/engine/server/scaling_handler.go]: 201 is used to indicate that "a scaling action has been created".  Since we don't observe whether it is actually scaled to the number that we expect.

and CF API PUT /v2/apps/:guid returns "201 created"

File: src/engine/server/scaling_handler_test.go (hunk 6 above)
Comment (boyang9527) [inline src/engine/server/scaling_handler_test.go]: will change to lagertest

File: src/cf/client.go (hunk 2 above)
Comment (fraenkel) [inline src/cf/client.go]: So I am not convinced that you don't have a race without always locking.

File: src/engine/server/scaling_handler.go (hunk 5 above)
Comment (fraenkel) [inline src/engine/server/scaling_handler.go]: which is just odd.
I prefer 200 which is what we discussed for the scheduler as well.

File: src/engine/server/server.go (hunk 7 above)
Comment (boyang9527) [inline src/engine/server/server.go]: I am trying to make the API a POST to a resource, not a verb/action.  I am good to change to /v1/scale, 

or maybe POST /v1/apps/{app-id}/scale

Comment (fraenkel) [inline src/engine/server/server.go]: Yes, that sounds better.

File: src/eventgenerator/aggregator/metricPoller_test.go (hunk 8 above)
Comment (boyang9527) [inline src/eventgenerator/aggregator/metricPoller_test.go]: I forgot to remove the commented line after testing. Will remove it

The ErrorResponse struct was moved to "models" package, eventgen  does not need to import from metricscollector any more

File: src/metricscollector/config/config.go (hunk 10 above)
Comment (boyang9527) [inline src/metricscollector/config/config.go]: It might be better to set time interval in seconds in the configuration file.  Do we want to have a finer grain?

Comment (fraenkel) [inline src/metricscollector/config/config.go]: Yes, for testing.
For the end user, they will only see seconds because we will generate the config.

File: src/cf/client.go (hunk 2 above)
Comment (boyang9527) [inline src/cf/client.go]: yes there is potential data race for  expire time. I will remove the outer check.

File: src/engine/server/scaling_handler.go (hunk 5 above)
Comment (boyang9527) [inline src/engine/server/scaling_handler.go]: ok.. sure I can change to 200

File: src/metricscollector/config/config.go (hunk 10 above)
Comment (boyang9527) [inline src/metricscollector/config/config.go]: ok, I will change back

**Comment by boyang9527** [comment]
@fraenkel  
I did another commit based on comments. Please review.

**Comment by fraenkel** [comment]
Merged with some modifications
